source/httpfs/driver.go

```diff
@@ -0,0 +1,159 @@
+package httpfs
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/golang-migrate/migrate/v4/source"
+)
+
+// driver is a migration source driver for reading migrations from
+// http.FileSystem instances. It implements source.Driver interface and can be
+// used as a migration source for the main migrate library.
+type driver struct {
+	migrations *source.Migrations
+	fs         http.FileSystem
+	path       string
+}
+
+// New creates a new migrate source driver from a http.FileSystem instance and a
+// relative path to migration files within the virtual FS. It is identical to
+// the WithInstance() function except it will delay any errors on fist usage of
+// this driver. This reduces the number of error handling branches in clients of
+// this package without compromising on correctness.
+func New(fs http.FileSystem, path string) source.Driver {
+	d, err := newDriver(fs, path)
+	if err != nil {
+		return &failedDriver{err}
+	}
+	return d
+}
+
+// WithInstance creates new migrate source driver from a http.FileSystem
+// instance and a relative path to migration files within the virtual FS.
+func WithInstance(fs http.FileSystem, path string) (source.Driver, error) {
+	return newDriver(fs, path)
+}
+
+func newDriver(fs http.FileSystem, path string) (*driver, error) {
+	root, err := fs.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	files, err := root.Readdir(0)
```

Comment:
Add a `Config` struct which specifies how many files to read at a time. The value will default to 0 for an "empty" struct, which will read all files in the directory.

Reply:
This is a tricky one.

We could start listing files in batches here. But if we still perform full directory scan in this function we are not gaining much. We will still use amount of memory proportional to the number of files in the directory because we are pushing them to `Migrations`. Are we optimizing for a case when the migration directory contain more non-migration files than migration ones? Are there any other advantages of batched directory scan that I am missing?

Substantial advantage would be if we could performing initial partial directory scan here, and continue scanning once we encounter a missing migration. But we could not do it since `Readdir()` returns files not in migration order. The order of files in directory will depend on OS and underlying FS. Performing partial initial scan we can not correctly compute First() migration number. Also for any cache miss we would still need a full scan of the directory to know for sure that the migration is really not present.

Reply:
Reading the directory in batch is useful for really slow FSs. e.g. remote sources like github, S3, etc.

Reply:
If we are optimizing for speed I do not think we will gain any. I mean if we are performing full directory scan in the `Init()` function it should take roughly the same amount of time (or less) to get directory listing in one call compared to multiple calls. We will still read full list and stat all files before returning from `Init()`. Am I missing something?

Reply:
If only some of the migrations are being run, there's no point in listing the whole directory.
e.g. if there are 100 migrations and you're currently at migration 50 and want to go to migration 75, there's no point in listing the last 25 migrations.

Reply:
But we still need to scan full directory just to know if some particular migration exists or not. This is all because directory contents ordering is not strictly defined. It depends on OS, file system or underlying transport layer.

It is easier to reason about it if we assume `Readdir()` returns file list in random order. Lets say we done one fixed length read for 100 elements from directory having 1000 entries:

* We do not know which files will we get in first `Readdir()` call.
* We do not know if we got migration with lowest number, and thus can not implement `First()` correctly.
* If we see gaps in migration sequences we do not know if migrations in the gaps exist or not, we can not implement `Next()` and `Prev()` correctly.
* We do not know which migrations will we get in the next `Readdir()` batch.

The only way to have any guarantees at all is to perform a complete directory scan. This means we need to do `root.Readdir(0)` or `for ... { ... root.Readdir(n) ... }` until we get `io.EOF`. In both cases we will traverse all files in the directory and both approaches should take roughly the same amount of time.

Reply:
[Docs](https://golang.org/pkg/os/#File.Readdir) for `Readdir()`:
> Readdir reads the contents of the directory associated with file and returns a slice of up to n FileInfo values, as would be returned by Lstat, in directory order. Subsequent calls on the same file will yield further FileInfos. 

Migrate currently assumes consistent ordering. The `create` command assumes that migrations are lexicographically sorted.

Reply:
The issue is that lexicographical ordering does not match directory order.

Example `main.go`:

```go
package main

import (
	"fmt"
	"io"
	"log"
	"net/http"
)

func main() {
	fs := http.Dir(".")

	d, err := fs.Open(".")
	if err != nil {
		log.Fatal(err)
	}

	for {
		fs, err := d.Readdir(2)
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Fatal(err)
		}
		for _, f := range fs {
			fmt.Println(f.Name())
		}
		fmt.Println("---")
	}
}
```

```sh
$ ls -l .
total 32
-rw-r--r--  1 folnin folnin     3 Dec 16 09:43 1000_migration.up.sql
-rw-r--r--  1 folnin folnin     3 Dec 16 09:43 1001_migration.up.sql
-rw-r--r--  1 folnin folnin     3 Dec 16 09:43 1002_migration.up.sql
-rw-r--r--  1 folnin folnin     3 Dec 16 09:43 1003_migration.up.sql
-rw-r--r--  1 folnin folnin     3 Dec 16 09:43 100_migration.up.sql
-rw-r--r--  1 folnin folnin     3 Dec 16 09:43 10_migration.up.sql
-rw-r--r--  1 folnin folnin     3 Dec 16 09:42 1_migration.up.sql
-rw-r--r--  1 folnin folnin   343 Dec 16 09:48 main.go
$ go run .
100_migration.up.sql
1_migration.up.sql
---
main.go
1001_migration.up.sql
---
1003_migration.up.sql
10_migration.up.sql
---
1002_migration.up.sql
1000_migration.up.sql
---
```

Command `ls` returns files in lexicographical ordering (which is not suitable for migration order as well as `1000_` < `1_`). Running main show the order in which files would be returned by iterative `Readdir()` calls (which is not lexicographical ant not migration order).

Reply:
Oh, looks like we'll need to read the whole directory and sort then...
The `create` command is designed to create migrations where the names are lexicographically sorted.

Reply:
That is the essence of the issue. Because we are forced to read full dir listing - there is not much being gained from reading file list in batches. We will be reading it in full anyway. As for sorting `ms.Append(m)` takes care of that.

Should we leave `Readdir(0)`?

Reply:
Yep, let's leave it as `Readdir(0)` and sort the results before using them. We can figure out a more clever solution in the future if necessary.